onnxruntime/core/providers/cpu/ml/treeregressor.cc

```diff
@@ -79,149 +60,220 @@ TreeEnsembleRegressor<T>::TreeEnsembleRegressor(const OpKernelInfo& info)
-      treeindex = val == threshold || tracktrue ? nodes_truenodeids_[treeindex] : nodes_falsenodeids_[treeindex];
-    } else if (mode == ::onnxruntime::ml::NODE_MODE::BRANCH_NEQ) {
-      treeindex = val != threshold || tracktrue ? nodes_truenodeids_[treeindex] : nodes_falsenodeids_[treeindex];
-    }
+TreeEnsembleRegressor<T>::~TreeEnsembleRegressor() {
+  delete[] nodes_;
+}
+
+#define TREE_FIND_VALUE(CMP)                                         \
+  if (has_missing_tracks_) {                                         \
+    while (root->mode != NODE_MODE::LEAF && loopcount >= 0) {        \
+      val = x_data[root->feature_id];                                \
+      root = (val CMP root->value ||                                 \
+              (root->missing_tracks == MissingTrack::TRUE &&         \
+               std::isnan(static_cast<T>(val))))                     \
+                 ? root->truenode                                    \
+                 : root->falsenode;                                  \
+      --loopcount;                                                   \
+    }                                                                \
+  } else {                                                           \
+    while (root->mode != NODE_MODE::LEAF && loopcount >= 0) {        \
+      val = x_data[root->feature_id];                                \
+      root = val CMP root->value ? root->truenode : root->falsenode; \
+      --loopcount;                                                   \
+    }                                                                \
+  }
 
-    if (treeindex < 0) {
-      return common::Status(common::ONNXRUNTIME, common::RUNTIME_EXCEPTION,
-                            "treeindex evaluated to a negative value, which should not happen.");
+template <typename T>
+common::Status TreeEnsembleRegressor<T>::ProcessTreeNode(float* predictions, TreeNodeElement* root,
+                                                         const T* x_data,
+                                                         unsigned char* has_predictions) const {
+  bool tracktrue;
+  T val;
+  if (same_mode_) {
+    int64_t loopcount = max_tree_depth_;
+    switch (root->mode) {
+      case NODE_MODE::BRANCH_LEQ:
+        TREE_FIND_VALUE(<=)
+        break;
+      case NODE_MODE::BRANCH_LT:
+        TREE_FIND_VALUE(<)
+        break;
+      case NODE_MODE::BRANCH_GTE:
+        TREE_FIND_VALUE(>=)
+        break;
+      case NODE_MODE::BRANCH_GT:
+        TREE_FIND_VALUE(>)
+        break;
+      case NODE_MODE::BRANCH_EQ:
+        TREE_FIND_VALUE(==)
+        break;
+      case NODE_MODE::BRANCH_NEQ:
+        TREE_FIND_VALUE(!=)
+        break;
+      case NODE_MODE::LEAF:
+        break;
+      default: {
+        std::ostringstream err_msg;
+        err_msg << "Invalid mode of value: " << static_cast<std::underlying_type<NODE_MODE>::type>(root->mode);
+        throw std::runtime_error(err_msg.str());
```

Comment:
>throw std::runtime_error(err_msg.str()); [](start = 8, length = 40)

Suggest using ORT_THROW, this would also provide string formatting.

Reply:
Done.